### PR TITLE
Random Sampling change - AVA-248

### PR DESCRIPTION
**Pull request by kroos47**

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -181,11 +181,13 @@ pub async fn do_main() -> Result<()> {
 		db.clone(),
 		ipfs.clone(),
 		cfg.max_parallel_fetch_tasks,
+		cfg.confidence,
```
Comment (prabal-banerjee) [inline src/main.rs:184]: Can we put some checks on the input value? For example, if someone supplies a negative confidence. Not necessarily inside `main`.

File: src/rpc.rs [hunk 2]
```diff
@@ -121,13 +121,13 @@ pub async fn get_block_by_number(url: &str, block: u64) -> Result<Block> {
 	}
 }
 
-pub fn generate_random_cells(max_rows: u16, max_cols: u16) -> Vec<Position> {
+pub fn generate_random_cells(max_rows: u16, max_cols: u16, cell_count: u32) -> Vec<Position> {
 	let max_cells = (max_rows as u32) * (max_cols as u32);
-	let count: u16 = if max_cells < 8 {
+	let count: u16 = if max_cells < cell_count as u32 {
 		// Multiplication cannot overflow since result is less than 8
 		max_rows * max_cols
```
Comment (prabal-banerjee) [inline src/rpc.rs:128]: Can we print a `debug` log when this case happens so that it is easier to debug later?

File: src/sync_client.rs [hunk 3]
```diff
@@ -65,7 +66,8 @@ async fn process_block(
 	let commitment = block_body.header.extrinsics_root.commitment;
 	let block_num = block_body.header.number;
 	// now this is in `u64`
-	let positions = rpc::generate_random_cells(max_rows, max_cols);
+	let cell_count = -((1f64 - confidence / 100f64).log2()).ceil() as u32;
```
Comment (prabal-banerjee) [inline src/sync_client.rs:69]: The formula appears in two places. For better maintainability, can we put it into a separate function and call from two places?

File: src/light_client.rs [hunk 4]
```diff
@@ -57,7 +58,13 @@ pub async fn run(
 					}
 					let commitment = header.extrinsics_root.commitment.clone();
 
-					let positions = rpc::generate_random_cells(max_rows, max_cols);
+					let cell_count: u32;
+					if confidence == 100f64 {
+						cell_count = -((1f64 - 99.9f64 / 100f64).log2()).ceil() as u32;
```
Comment (prabal-banerjee) [inline src/light_client.rs:63]: Should we increase precision in this case? Or do we want a conservative default?

File: src/light_client.rs [hunk 5]
```diff
@@ -57,7 +58,13 @@ pub async fn run(
...
+					} else {
+						cell_count = (-((1f64 - confidence / 100f64).log2())).ceil() as u32;
```
Comment (prabal-banerjee) [inline src/light_client.rs:65]: Can we write test cases to ascertain correctness? For example, giving `confidence=60` may give output `1` whereas `1` sample can only give atmost `50%` confidence.

**Review by prabal-banerjee — CHANGES_REQUESTED** [review]

File: src/light_client.rs (hunk 5 above)
Comment (kroos47) [reply src/light_client.rs:65]: yeh sure! But how its being checked is if we put 60 as confidence then it checks for number on cells which gives >=60.

File: src/light_client.rs [hunk 6]
```diff
@@ -57,7 +58,13 @@ pub async fn run(
...
+					}
```
Comment (aterentic-ethernal) [inline src/light_client.rs:66]: Can we also move code to separate function, e.g. `cell_count_for_confidence` or something better?

Comment (aterentic-ethernal) [reply src/light_client.rs:66]: Actually, already mentioned :D

File: src/light_client.rs (hunk 5 above)
Comment (prabal-banerjee) [reply src/light_client.rs:65]: I don't think `cell_count` of `1` gives `>=60` confidence.

File: src/rpc.rs [hunk 7]
```diff
@@ -121,13 +121,18 @@ pub async fn get_block_by_number(url: &str, block: u64) -> Result<Block> {
 	}
 }
 
-pub fn generate_random_cells(max_rows: u16, max_cols: u16) -> Vec<Position> {
+pub fn generate_random_cells(max_rows: u16, max_cols: u16, cell_count: u32) -> Vec<Position> {
 	let max_cells = (max_rows as u32) * (max_cols as u32);
-	let count: u16 = if max_cells < 8 {
+	let count: u16 = if max_cells < cell_count as u32 {
 		// Multiplication cannot overflow since result is less than 8
```
Comment (sh3ll3x3c) [inline src/rpc.rs:127]: Should we remove this comment (not relevant anymore)?

File: src/rpc.rs [hunk 8]
```diff
@@ -121,13 +121,18 @@ pub async fn get_block_by_number(url: &str, block: u64) -> Result<Block> {
 	}
 }
 
-pub fn generate_random_cells(max_rows: u16, max_cols: u16) -> Vec<Position> {
+pub fn generate_random_cells(max_rows: u16, max_cols: u16, cell_count: u32) -> Vec<Position> {
```
Comment (sh3ll3x3c) [inline src/rpc.rs:124]: Do we need some sort of sanity check for the cell count, ie in order not to overflow the multiplication?

Comment (kroos47) [reply src/rpc.rs:124]: yeh this needs to be done!!

**Comment by kroos47** [comment]
opening a rebased one

**Pull request closed without merging**